# Introduction to Conditional Probability (Worksheet)

In [1]:
import pandas as pd
df = pd.read_csv('TIA_1987_2016_with_dates.csv', index_col=0, parse_dates=True)

'''
Name: Riley Zuckert
Date: 1/21/22
ISTA 331
Collaborator(s):
'''

'\nName: Riley Zuckert\nDate: 1/21/22\nISTA 331\nCollaborator(s):\n'

## Part 1: Probability calculations

1. In the entire data frame, what is the frequency of the event that the high temperature is over 90 degrees? What is the estimated probability that the high temperature is over 90 degrees?

In [6]:
B = df[df.MaxT > 90]
freq = len(B)
freq

4520

In [7]:
freq / len(df)

0.4124840299324694

The frequency of events where the high temperature is over 90 degrees is 4,520 days. 
The estimated probability that the high temperature is over 90 degrees is approximately 41.25%.

2. What is the estimated probability that a day is in June, July, or August? 

* Hint 1: the `DatetimeIndex` type that is used here has an attribute `month`. Try evaluating `df.index.month` to see how it works. You can use this to create a Boolean expression to index out the months you want.
* Hint 2: as a consistency check, this should be pretty close to 25%. Why is this true?

In [39]:
B = df[df.index.month.isin([6,7,8])]
len(B) / len(df)

0.25187077933929547

The estimated probability that a day is in June, July, or August is approximately 25.19%. This number is close to 25% as the three of these months make up 1/4 of the calendar, or 25% of the months of the year.

3. What is the estimated probability that a day has a high temperature over 90 degrees **and** is in June, July, or August?

In [42]:
len(df[(df.MaxT > 90) & (df.index.month.isin([6,7,8]))]) / len(df)

0.23854718014236173

The estimated probability that a day has a high temperature over 90 degrees and is in June, July, or August is approximately 23.85%.

4. What is the estimated probability that a day has a high temperature over 90 degrees **given** it is in June, July, or August?

In [44]:
B = df[df.index.month.isin([6,7,8])]
A_and_B = B[B.MaxT > 90]
P_A_given_B = len(A_and_B) / len(B)
P_A_given_B

0.9471014492753623

The estimated probability that a day has a high temperature over 90 degrees given it is in June, July, or August is approximately 94.71%.

5. What is the estimated probability that a day is in June, July, or August given its high temperature is over 90 degrees? Is your answer the same as the answer to part 4?

In [46]:
B = df[df.MaxT > 90]
A_and_B = B[B.index.month.isin([6,7,8])]
P_A_given_B = len(A_and_B) / len(B)
P_A_given_B

0.5783185840707965

The estimated probability that a day is in June, July, or August given its high temperature is over 90 degrees is approximately 57.83%.
No, this answer is not the same as the answer to part 4 because P(A) != P(B).

## Part 2: Function writing

1. Write a function called `get_P` that takes a frame, an attribute name (as a string), a cutoff, another attribute name, and another cutoff. It should return the conditional probability that the first attribute is greater than the first cutoff given that the second attribute is greater than the second cutoff.

In [52]:
def get_P(frame, attribute_1, cutoff_1, attribute_2, cutoff_2):
    B = frame[frame[attribute_2] > cutoff_2]
    A_and_B = B[B[attribute_1] > cutoff_1]
    P_A_given_B = len(A_and_B) / len(B)
    return P_A_given_B

In [53]:
### Run this cell to test your code
from math import isclose
correct_1 = 0.8011734028683182
correct_2 = 0.11490125673249552

assert(isclose(correct_1, get_P(df, 'MaxT', 70, 'MinT', 30)))
assert(isclose(correct_2, get_P(df, 'Pcpn', 0, 'MaxT', 100)))

In [54]:
get_P(df, 'MaxT', 70, 'MinT', 30)

0.8011734028683182

In [55]:
get_P(df, 'Pcpn', 0, 'MaxT', 100)

0.11490125673249552